polysh/display_names.py

```diff
@@ -16,73 +16,71 @@
+# Used for terminal sizes and layouting
 max_display_name_length = 0
 
 
-class DisplayNamePrefix(object):
-    def __init__(self):
-        self.next_suffix = 0
-        self.holes = RBTree()
+def acquire_prefix_index(prefix):
+    if prefix not in PREFIXES:
+        PREFIXES[prefix] = list()
 
-    def new_suffix(self):
-        if len(self.holes) == 0:
-            suffix = self.next_suffix
-            self.next_suffix += 1
-        else:
-            first_node = self.holes.firstNode()
-            suffix = first_node.key
-            self.holes.deleteNode(first_node)
-        return suffix
-
-    def putback_suffix(self, suffix):
-        if suffix + 1 != self.next_suffix:
-            self.holes.insertNode(suffix, suffix)
-            return
+    # Search and reuse removed host suffix
+    for idx, item in enumerate(PREFIXES[prefix]):
+        if not item:
+            PREFIXES[prefix][idx] = True
+            return idx
+
+    # Add new suffix if no old suffix can be reused
+    PREFIXES[prefix].append(True)
+    return len(PREFIXES[prefix]) -1
 
-        self.next_suffix = suffix
-        while True:
-            prev_suffix = self.next_suffix - 1
-            prev_suffix_node = self.holes.findNode(prev_suffix)
-            if not prev_suffix_node:
-                return
-            self.holes.deleteNode(prev_suffix_node)
-            self.next_suffix = prev_suffix
 
-    def empty(self):
-        return self.next_suffix == 0
+def release_prefix_index(prev_display_name):
+    split = prev_display_name.split('#')
+    prefix = split[0]
+    if len(split) == 1:
+        suffix = 0
+    else:
+        suffix = int(split[1])
 
+    # We are not deleting the host with the highest suffix. Therefore we need
+    # to mark the current suffix index as unused.
+    if suffix != len(PREFIXES[prefix]) -1:
+        PREFIXES[prefix][suffix] = False
+        return
 
-def make_unique_name(prefix):
-    prefix_obj = PREFIXES.get(prefix, None)
-    if prefix_obj is None:
-        prefix_obj = DisplayNamePrefix()
-        PREFIXES[prefix] = prefix_obj
+    # We are deleting the host with thte highest suffix. Therefore we need to
+    # delete it.
+    PREFIXES[prefix].pop(suffix)
+
+    # Remove holes previously left.
+    for idx in reversed(range(len(PREFIXES[prefix]))):
+        if PREFIXES[prefix][idx]:
+            return
+        PREFIXES[prefix].pop(idx)
 
-    suffix = prefix_obj.new_suffix()
+    # If we arrived here, we just deleted the last item with a specific prefix. Therefore we need to delete the whole prefix now.
+    del PREFIXES[prefix]
+
+
+def make_unique_name(prefix):
+    suffix = acquire_prefix_index(prefix)
     if suffix:
-        name = '%s#%d' % (prefix, suffix)
+        return '{}#{}'.format(prefix, suffix)
     else:
-        name = prefix
-
-    return name
+        return prefix
 
 
 def update_max_display_name_length():
     from polysh import dispatchers
```

Comment:
Maybe this can be imported properly in the beginning now.

Reply:
Moving it works in python3.5 and python3.6 but not python3.4
https://travis-ci.org/innogames/polysh/jobs/388869052